ragna/assistants/_openai.py

```diff
@@ -23,9 +23,45 @@ def _make_system_content(self, sources: list[Source]) -> str:
         )
         return instruction + "\n\n".join(source.content for source in sources)
 
-    def _stream(
-        self, prompt: str, sources: list[Source], *, max_new_tokens: int
+    def _render_prompt(self, prompt: Union[str,list[Message]], system_prompt: str) -> list[dict]:
+        """
+        Ingests ragna messages-list or a single string prompt and converts to assistant-appropriate format.
+
+        Returns:
+            ordered list of dicts with 'content' and 'role' keys
+        """
+        if isinstance(prompt,str):
+            messages = [
+                        {
+                            "role": "system",
+                            "content": system_prompt,
+                        },
+                        {
+                            "role": "user",
+                            "content": prompt,
+                        },
+                    ]
+            return messages
+        else:
+            system_message = [{"role":"system", "content":system_prompt}]
+            messages = [{"role":i["role"],"content":i["content"]} for i in prompt if i["role"] != "system"]
+            return system_message.extend(messages)
+    
+    async def generate(
+        self, prompt: Union[str,list[Message]], system_prompt: str, *, max_new_tokens: int
```

Comment:
Both `system_prompt` and `max_new_tokens` should have a reasonable default values. We don't want to force users to always put something there. Plus, `system_prompt` should probably also be keyword only. Calling `.generate()` with two strings filled with instructions can become confusing real quick.